Neos.Neos/Classes/AssetUsage/Projection/AssetUsageProjection.php

```diff
@@ -200,15 +199,14 @@ private function getAssetIdsByProperty(SerializedPropertyValues $propertyValues)
     }
 
     /**
-     * @param string $type
-     * @param mixed $value
+     * @param int|float|string|bool|array<int|string,mixed>|\ArrayObject<int|string,mixed>|null $value
      * @return array<string>
      * @throws InvalidTypeException
      */
-    private function extractAssetIds(string $type, mixed $value): array
+    private function extractAssetIds(string $type, int|float|string|bool|array|\ArrayObject|null $value): array
```

Comment:
This union type seems somewhat "random". Are we sure that it is exhaustive?

Reply:
It's the union type of SerializedPropertyValue, so not random, but prone to errors on change I guess. I don't know why the SerializedPropertyValue is not passed instead

Reply:
> It's the union type of SerializedPropertyValue, so not random

Ah, right

> I don't know why the SerializedPropertyValue is not passed instead

Because this is called recursively.. And that's why it had the `mixed` type hint.

I'm not sure whether a `SerializedPropertyValue` can be an array of something that is _not_ in the union type. But the fact that I would have to guess makes this part fragile.

Is this maybe better:
```php
/**
 * @return array<string>
 * @throws InvalidTypeException
 */
private function extractAssetIds(SerializedPropertyValue $propertyValue): array
{
    if (is_string($propertyValue->value)) {
        preg_match_all('/asset:\/\/(?<assetId>[\w-]*)/i', $propertyValue->value, $matches, PREG_SET_ORDER);
        return array_map(static fn(array $match) => $match['assetId'], $matches);
    }
    if (is_subclass_of($propertyValue->type, ResourceBasedInterface::class)) {
        return isset($value['__identifier']) ? [$value['__identifier']] : [];
    }

    // Collection type?
    /** @var array{type: string, elementType: string|null, nullable: bool} $parsedType */
    $parsedType = TypeHandling::parseType($propertyValue->type);
    if ($parsedType['elementType'] === null) {
        return [];
    }
    if (
        !is_subclass_of($parsedType['elementType'], ResourceBasedInterface::class)
        && !is_subclass_of($parsedType['elementType'], \Stringable::class)
    ) {
        return [];
    }
    if (!is_iterable($value)) {
        return [];
    }
    $assetIds = [];
    foreach ($value as $elementValue) {
        $assetIds[] = $this->extractAssetIds(new SerializedPropertyValue($elementValue, $parsedType['elementType']));
    }
    return array_merge(...$assetIds);
}
```

or maybe even a version [without recursion](https://www.refactoring.com/catalog/replaceRecursionWithIteration.html)?

Reply:
ah, you're right, I forgot the recursion.
I'm not sure either about what can be in a SerializedPropertyValue (would it make sense for a SerializedPropertyValue to consist only of primitives and other SerializedPropertyValues?), so I'd be fine with mixed as well for now and let the tests handle this instead of the static analyzer.
I think it would make sense to revisit this later and would be happy if we got this merged before :)